x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java

```diff
@@ -0,0 +1,280 @@
+import org.elasticsearch.xpack.esql.type.EsqlDataTypes;
+import org.elasticsearch.xpack.ql.expression.Expression;
+import org.elasticsearch.xpack.ql.expression.Nullability;
+import org.elasticsearch.xpack.ql.tree.NodeInfo;
+import org.elasticsearch.xpack.ql.tree.Source;
+import org.elasticsearch.xpack.ql.type.DataType;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Function;
+
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.ParamOrdinal.FIRST;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.ParamOrdinal.SECOND;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isType;
+
+/**
+ * Appends values to a multi-value
+ */
+public class MvAppend extends EsqlScalarFunction implements EvaluatorMapper {
+    private final Expression field1, field2;
+    private DataType dataType;
+
+    @FunctionInfo(
+        returnType = {
+            "boolean",
+            "cartesian_point",
+            "cartesian_shape",
+            "date",
+            "double",
+            "geo_point",
+            "geo_shape",
+            "integer",
+            "ip",
+            "keyword",
+            "long",
+            "text",
+            "version" },
+        description = "Concatenates values of two multi-value fields."
+    )
+    public MvAppend(
+        Source source,
+        @Param(
+            name = "field1",
+            type = {
+                "boolean",
+                "cartesian_point",
+                "cartesian_shape",
+                "date",
+                "double",
+                "geo_point",
+                "geo_shape",
+                "integer",
+                "ip",
+                "keyword",
+                "long",
+                "text",
+                "version" }
+        ) Expression field1,
+        @Param(
+            name = "field2",
+            type = {
+                "boolean",
+                "cartesian_point",
+                "cartesian_shape",
+                "date",
+                "double",
+                "geo_point",
+                "geo_shape",
+                "integer",
+                "ip",
+                "keyword",
+                "long",
+                "text",
+                "version" }
+        ) Expression field2
+    ) {
+        super(source, Arrays.asList(field1, field2));
+        this.field1 = field1;
+        this.field2 = field2;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        TypeResolution resolution = isType(field1, EsqlDataTypes::isRepresentable, sourceText(), FIRST, "representable");
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+        dataType = field1.dataType();
+        return isType(field2, t -> t == dataType, sourceText(), SECOND, dataType.typeName());
```

Comment:
```suggestion
        return isType(field2, t -> t == NULL || t == dataType, sourceText(), SECOND, dataType.typeName());
```

Otherwise `MV_APPEND(some_type, null)` will work, but MV_APPEND(null, some_type) won't. Might be worth adding some tests with this too.

Reply:
> Checking isType(), it already checks for NULL from what I see: predicate.test(e.dataType()) || e.dataType() == NULL.

In this case the first type, "t" / `dataType`, needs to be checked against `NULL`, rather than `field2`'s type (which is `e` in `isType()`).

>  if any field is null, set dataType to NULL. 

That's correct (and the code does it right also for the `MV_APPEND(null, some_type)`, case). 
But in one case the `null` result has first argument's type, while in the other `null` has the `NULL` type.